src/net/socket_state.rs

```diff
@@ -40,30 +42,57 @@ impl SocketState {
     }
 
     /// This will initialize the seq number, ack number and give back the raw data of the packet with the updated information.
-    pub fn pre_process_packet(&mut self, packet: Packet) -> Result<(SocketAddr, Vec<u8>)> {
+    pub fn pre_process_packet(&mut self, packet: Packet, config: &NetworkConfig) ->  Result<(SocketAddr, PacketData)>  {
         let connection = self.create_connection_if_not_exists(&packet.addr)?;
-        // queue new packet
-        if let Ok(mut lock) = connection.write() {
-            let seq_num = lock.seq_num;
-            lock.waiting_packets.enqueue(seq_num, packet.clone());
-        }
 
-        let raw_packet: RawPacket;
-        // initialize packet data, seq, acked_seq etc.
-        let mut l = connection
+        let mut lock = connection
             .write()
             .map_err(|_| NetworkError::AddConnectionToManagerFailed)?;
 
-        raw_packet = RawPacket::new(
-            l.seq_num,
-            &packet,
-            l.their_acks.last_seq,
-            l.their_acks.field,
-        );
-        // increase sequence number
-        l.seq_num = l.seq_num.wrapping_add(1);
-        let buffer = serialize(&raw_packet)?;
-        Ok((packet.addr, buffer))
+        let connection_seq = lock.seq_num;
+        // queue new packet
+        lock.waiting_packets.enqueue(connection_seq, packet.clone());
+
+        let mut packet_data = PacketData::new();
+
+        // create packet header
+        let packet_header = PacketHeader::new(connection_seq, lock.their_acks.last_seq, lock.their_acks.field);
+
+
+        let payload = packet.payload();
+        let payload_length = payload.len();
+
+        if payload_length <= config.fragment_size {
+            // we don't need to splitup the packet packet
+            packet_data.add_fragment(&packet_header, payload.to_vec());
+        }else {
+            // we need to split up the packet.
+            // check how many fragments this packet should be seperated in.
+            let remainder = if payload_length % config.fragment_size > 0 { 1 } else { 0 };
+            let num_fragments = ((payload_length / config.fragment_size) + remainder) as usize;
+
+            for fragment_id in 0..num_fragments{
+                let fragment = FragmentHeader::new(fragment_id as u8, num_fragments as u8, packet_header.clone());
```

Comment:
Can we ensure this downcasting will work? Going from `usize` to `u8`?

Reply:
For a certain part. Mostly you don't want to have more than 256 fragments because the change for broken packet will be bigger.

If you would have a payload of +- 270.000 bytes this could go wrong indeed if you do the math. This would not happen probably but going to see if I could make it safer!